stage1/init/common/units.go

```diff
@@ -195,6 +182,150 @@ func ImmutableEnv(p *stage1commontypes.Pod) error {
 	return uw.Error()
 }
 
+// SetupAppIO prepares all properties related to streams (stdin/stdout/stderr) and TTY
+// for an application service unit.
+//
+// It works according to the following steps:
+//  1. short-circuit interactive pods and legacy systemd, for backward compatibility
+//  2. parse app-level annotations to determine stdin/stdout/stderr mode
+//     2a. if an annotation is missing/invalid, it fallbacks to legacy mode (in: null, out/err: journald)
+//     2b. if a valid annotation is found, it prepares:
+//          - TTY and stream properties for the systemd service unit
+//          - env variables for iottymux binary
+//  3. if any of stdin/stdout/stderr is in TTY or streaming mode:
+//     3a. the env file for iottymux is written to `/rkt/iottymux/<appname>/env` with the above content
+//     3b. for TTY mode, a `TTYPath` property and an `After=ttymux@<appname>.service` dependency are added
+//     3c. for streaming mode, a `Before=iomux@<appname>.service` dependency is added
+//
+// For complete details, see dev-docs at Documentation/devel/log-attach-design.md
+func (uw *UnitWriter) SetupAppIO(p *stage1commontypes.Pod, ra *schema.RuntimeApp, binPath string, opts ...*unit.UnitOption) []*unit.UnitOption {
+	if uw.err != nil {
+		return opts
+	}
+
+	if p.Interactive {
+		opts = append(opts, unit.NewUnitOption("Service", "StandardInput", "tty"))
+		opts = append(opts, unit.NewUnitOption("Service", "StandardOutput", "tty"))
+		opts = append(opts, unit.NewUnitOption("Service", "StandardError", "tty"))
+		return opts
+	}
+
+	flavor, systemdVersion, err := GetFlavor(uw.p)
+	if err != nil {
+		uw.err = err
+		return opts
+	}
+	// Attach needs https://github.com/systemd/systemd/pull/4179, ie. systemd v232 or a backport
+	if ((flavor == "src" || flavor == "host") && systemdVersion < 232) ||
```

Comment:
We probably don't gain much from that, but I prefer explicit error as a general approach. Added it.